### PR TITLE
Add support write file custom metadata

### PR DESCRIPTION
**Pull request by eisberg**
Added support for saving custom metadata

See in parquet-dotnet implementation:
https://github.com/aloneguid/parquet-dotnet/blob/dd88943c900c7da2afc20f4bbd8466aeeae23100/docs/writing.md#custom-metadata

### CODE REVIEW

**Comment by Katalystical** [comment]
Thanks for contributing, this seems like a nice addition.
I'd like to suggest not relying on KeyValue-objects/array being required by the setCustomMetadata function (or return values from getCustomMetadata) in favor of simply using PHP's associative arrays and do the AssocArray2KeyValue conversion internally.
We'd have to make sure the keys/values are strings, though and cover some edge cases (null/empty strings, etc.) in tests.

And of course, we really need some integration testing, if this is fully parsed by other Parquet implementations. Ideally, there's a test file produced by a different implementation we could read for testing purposes.

**Comment by eisberg** [comment]
@Katalystical 

Redesigned the way to set metadata: associated array and added the **metadata** property of the schema object
 by analogy as it is implemented in pyarrow. Added a test for writing/reading metadata, including reading a file saved in pyarrow:



  ```python
import pyarrow.csv as pv
import pyarrow.parquet as pq
import pyarrow as pa

  # Write:
    table = pv.read_csv('movies.csv')

    my_schema = pa.schema([
        pa.field("movie", "string", False),
        pa.field("release_year", "int64", True)],
        metadata={"great_music": "reggaeton"})
    t2 = table.cast(my_schema)

    pq.write_table(t2, './movies.parquet')

    # Read:
    s = pq.read_table('movies.parquet').schema
    print(s.metadata[b'great_music']) # => b'reggaeton'
```

php:

```php

    // Write:
    $ms = fopen('php://memory', 'r+');

    $id = DataField::createFromType('id', 'integer');
    $metadata = [
      'great_music'=>'reggaeton',
    ];

    $writer = new ParquetWriter(new Schema([$id], $metadata), $ms);

    $rg = $writer->CreateRowGroup();
    $rg->WriteColumn(new DataColumn($id, [ 1, 2, 3, 4 ]));
    $rg->finish();
    $writer->finish();

    // Read:
    $schema = (new ParquetReader($ms))->schema;
    echo $schema->metadata['great_music']; // reggaeton
```

File: src/format/FileMetaData.php [hunk 1]
```diff
@@ -207,6 +209,57 @@ public function getName()
     }
 
 
+  private function convertAssocToKeyValues(array $customMeta): array{
```
Comment (Katalystical) [inline src/format/FileMetaData.php]: This is an auto-generated file from the Thrift sources. Please don't modify, as it will be overwritten during a re-generation.

**Comment by Katalystical** [comment]
Thanks again for contributing.
I'm fine doing the assoc<->KeyValue conversion right in ParquetWriter. I found your initial solution pretty easy to understand and minimalistic, let's do it this way, but doing the mapping to KeyValue[] internally (so, let it accept regular PHP assoc arrays).

Btw. I'm unsure about the `@param array<string, string>` PHPDoc generics-style arrays. Seems this is not accepted in PSR-5 and won't be supported by most IDEs/editors. It's simply preliminary.

**Comment by eisberg** [comment]
Ok, I returned the implementation to the ParquetWriter::setCustomMetadata/ParquetReader::getCustomMetadata

**Comment by eisberg** [comment]
@Katalystical  now is ok?

**Comment by Katalystical** [comment]
Merged into develop branch: https://github.com/codename-hub/php-parquet/pull/14
Sorry for the late response, been busy.

**Comment by Katalystical** [comment]
@eisberg I've improved your implementation. 
See https://github.com/codename-hub/php-parquet/commit/77d6dd97e0b25aa8b9aedb963d9f4bada1312389
Seems NULLs and empty strings **are allowed** by the spec (see Thrift and real implementations, f.e. Rust) which defines the **value** to be Optional, therefore: being _NULL_ or an _empty string_ is ok. Even an empty string as key is possible ;)
Added some test cases to it, still to be integration-tested with other Parquet Implementations that can also read (and probably write) custom metadata.
